### PR TITLE
RavenDB-12421 unbounded compaction in 32-bit mode could sometime cras…

### PR DESCRIPTION
**Pull request by talweiss1982**
…h the storage

### CODE REVIEW

File: src/Raven.Server/Documents/TransactionOperationsMerger.cs [hunk 1]
```diff
@@ -826,7 +827,7 @@ private enum PendingOperations
                 var modifiedSize = llt.NumberOfModifiedPages * Constants.Storage.PageSize;
 
                 var canCloseCurrentTx = previousOperation == null || previousOperation.IsCompleted;
-                if (canCloseCurrentTx)
+                if (canCloseCurrentTx || llt.Environment.Options.ForceUsing32BitsPager || PlatformDetails.Is32Bits)
```
Comment (ayende) [inline src/Raven.Server/Documents/TransactionOperationsMerger.cs]: Can we extract this computation to the constructor? 
I want to make it easy to ignore

Comment (ayende) [reply src/Raven.Server/Documents/TransactionOperationsMerger.cs]: Also, what about the behavior a bit below, where we do the check again in:

```
// if I can't close the tx, this means that the previous async operation is still in progress
// and there are incoming requests coming in. We'll accept them, to a certain limit
if (modifiedSize < _maxTxSizeInBytes)
    continue; // we can still process requests at this time, so let's do that...
```

File: src/Voron/Impl/Paging/Windows32BitsMemoryMapPager.cs [hunk 2]
```diff
@@ -338,9 +339,18 @@ private LoadedPage MapPages(TransactionState state, long startPage, long size)
...
+
+                    const int ERROR_NOT_ENOUGH_MEMORY = 8;
+                    if (lastWin32Error == ERROR_NOT_ENOUGH_MEMORY)
+                    {
+                        throw new OutOfMemoryException($"Unable to map {size / Constants.Size.Kilobyte:#,#0} kb starting at {startPage} on {FileName}", new Win32Exception(lastWin32Error));
+                    }
+                    else
+                    {
+                        throw new Win32Exception(
+                            $"Unable to map {size / Constants.Size.Kilobyte:#,#0} kb starting at {startPage} on {FileName}",
```
Comment (ayende) [inline src/Voron/Impl/Paging/Windows32BitsMemoryMapPager.cs]: Include the `lastWin32Error ` here too

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Documents/TransactionOperationsMerger.cs (hunk 1 above)
Comment (talweiss1982) [reply src/Raven.Server/Documents/TransactionOperationsMerger.cs]: I'm not sure what should be the behavior in the merger, I don't want to start rejecting requests maybe it is better if @aviviadi  or @haludi will look into that?
_maxTxSizeInBytes controls the max transaction size it is set to 4MB for 32 bit
I'm not sure what the behavior is if we break from this loop...

Comment (talweiss1982) [reply src/Raven.Server/Documents/TransactionOperationsMerger.cs]: actually the merger is limiting this to 4MB for 32bit so we can remove this entirly

File: src/Raven.Server/Documents/TransactionOperationsMerger.cs [hunk 3]
```diff
@@ -67,6 +68,7 @@ public TransactionOperationsMerger(DocumentDatabase parent, CancellationToken sh
 
         public void Start()
         {
+            Is32Bits = _parent.DocumentsStorage.Environment.Options.ForceUsing32BitsPager || PlatformDetails.Is32Bits;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/TransactionOperationsMerger.cs]: We need to move this to ctor and mark it as readonly

Comment (ppekrol) [reply src/Raven.Server/Documents/TransactionOperationsMerger.cs]: You can do that by accessing database.Configuration.Storage.ForceUsing32BitsPager

**Review by ppekrol — APPROVED** [review]

**Pull request merged**